File: 5. Recursion I/509. Fibonacci Number/fibonacci.cpp
class Solution
{
public:
    unordered_map<int, int> cache;

    int fib(int n)
    {
        if (cache.find(n) != cache.end())
        {
            return cache[n];
        }

        if (n < 2)
        {
            return n;
        }

        int result;
        result = fib(n - 2) + fib(n - 1);
        cache[n] = result;

        return result;
    }
};

File: 5. Recursion I/509. Fibonacci Number/fibonacci.py
class Solution:
    def __init__(self):
        self.cache = {}

    def fib(self, n: int) -> int:
        if n in self.cache:
            return self.cache[n]
        
        if n < 2:
            return n

        result = self.fib(n-1) + self.fib(n-2)
        self.cache[n] = result

        return result
